src/kernels/jupyter/launcher/serverUriStorage.ts

```diff
@@ -140,58 +142,56 @@ export class JupyterServerUriStorage implements IJupyterServerUriStorage {
             undefined
         );
     }
-    public getUri(): Promise<string> {
+    private getUris(): Promise<string[]> {
         if (!this.currentUriPromise) {
             this.currentUriPromise = this.getUriInternal();
         }
 
         return this.currentUriPromise;
     }
-    public async getRemoteUri(): Promise<string | undefined> {
-        const uri = await this.getUri();
-        switch (uri) {
-            case Settings.JupyterServerLocalLaunch:
-                return;
-            case Settings.JupyterServerRemoteLaunch:
-                // In `getUriInternal` its not possible for us to end up with Settings.JupyterServerRemoteLaunch.
-                // If we do, then this means the uri was never saved or not in encrypted store, hence no point returning and invalid entry.
-                return;
-            default:
-                return uri;
-        }
+    public async getRemoteUris(): Promise<string[]> {
+        return this.getUris();
     }
     public async setUriToLocal(): Promise<void> {
-        await this.setUri(Settings.JupyterServerLocalLaunch);
-    }
-    public async setUriToRemote(uri: string, displayName: string): Promise<void> {
-        await this.setUri(uri);
-        await this.addToUriList(uri, Date.now(), displayName);
+        this.remoteUris = [];
+        this.currentUriPromise = Promise.resolve(this.remoteUris);
+        await this.serverConnectionType.setIsLocalLaunch(true);
+        this._onDidChangeUri.fire();
     }
-
-    public async setUri(uri: string) {
+    public async addRemoteUri(uri: string, displayName: string): Promise<void> {
+        this.remoteUris.push(uri);
         // Set the URI as our current state
-        this.currentUriPromise = Promise.resolve(uri);
-        if (uri === Settings.JupyterServerLocalLaunch) {
-            await this.serverConnectionType.setIsLocalLaunch(true);
-        } else {
-            await this.addToUriList(uri, Date.now(), uri);
-            await this.serverConnectionType.setIsLocalLaunch(false);
+        this.currentUriPromise = Promise.resolve(this.remoteUris);
```

Comment:
Does just resolving the promise here make sense now that we are adding to a list as opposed to just setting?

Reply:
Trying to trace it now. getUriInternal might be in progress to fetch the stored URIs, but that could totally be replaced by this resolve here, wouldn't you want a combination of the stored fetch + the new value?

Reply:
we always end up getting all of the uris in the ctor.
Hence we'll alraedy have the stuff from cache, but will update to await on that promise first.